public/app/features/query/components/QueryGroup.tsx

```diff
@@ -290,6 +279,42 @@ export class QueryGroup extends PureComponent<Props, State> {
     );
   };
 
+  renderDataSourcePickerWithPrompt = () => {
+    const { isDataSourceModalOpen } = this.state;
+
+    const commonProps = {
+      enableFileUpload: config.featureToggles.editPanelCSVDragAndDrop,
+      fileUploadOptions: {
+        onDrop: this.onFileDrop,
+        maxSize: DFImport.maxFileSize,
+        multiple: false,
+        accept: DFImport.acceptedFiles,
+      },
+      current: this.props.options.dataSource,
+      onChange: (ds: DataSourceInstanceSettings) => {
+        this.onChangeDataSource(ds);
+        this.setState({ isDataSourceModalOpen: false });
+      },
+    };
+    const onDismiss = () => this.setState({ isDataSourceModalOpen: false });
+
+    return (
+      <>
+        {isDataSourceModalOpen && config.featureToggles.advancedDataSourcePicker && (
+          <DataSourceModal {...commonProps} onDismiss={onDismiss}></DataSourceModal>
+        )}
+        <DataSourcePicker
+          {...commonProps}
+          metrics={true}
+          mixed={true}
+          dashboard={true}
+          variables={true}
+          onClickAddCSV={this.onClickAddCSV}
+        />
+      </>
+    );
+  };
```

Comment:
this feels to me like the wrong place for this logic, the showing of the modal and picking a data source for the first panel should be part of a "add the first panel" workflow that should be separate from code that deals with the edit scenario. 

Just trying to think of how to approach this from "single responsibility" principle perspective and how to avoid components that handle to many scenarios and workflows.